metrics/density/footprint_data.sh

```diff
@@ -0,0 +1,354 @@
+	fi
+
+	# Grabbing slab total from meminfo is easier than doing the math
+	# on slabinfo
+	item=$(cat /proc/meminfo | fgrep "Slab:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+}
+
+function get_memfree() {
+	mem_free=$(sudo smem -w | head -6 | tail -1 | awk '{print $4}')
+	mem_free=$(($mem_free * 1024))
+	echo $mem_free
+}
+
+function grab_system() {
+	if [[ $DUMP_HEADER ]]; then
+		echo -n ", avail" >> $DATAFILE
+		echo -n ", avail_incr" >> $DATAFILE
+		echo -n ", cached" >> $DATAFILE
+		echo -n ", free_smem" >> $DATAFILE
+		echo -n ", free_smem_incr" >> $DATAFILE
+		echo -n ", anon_pages" >> $DATAFILE
+		echo -n ", mapped_pages" >> $DATAFILE
+		echo -n ", cached" >> $DATAFILE
+
+		# Store up baseline measures
+		base_mem_avail=$(free -b | head -2 | tail -1 | awk '{print $7}')
+		base_mem_free=$(get_memfree)
+		return
+	fi
+
+	# avail memory, from 'free'
+	item=$(free -b | head -2 | tail -1 | awk '{print $7}')
+	echo -n ", $item" >> $DATAFILE
+	avail_incr=$(($base_mem_avail-$item))
+	echo -n ", $avail_incr" >> $DATAFILE
+
+	# cached memory, from 'free'
+	item=$(free -b | head -2 | tail -1 | awk '{print $6}')
+	echo -n ", $item" >> $DATAFILE
+
+	# free memory from smem
+	item=$(get_memfree)
+	echo -n ", $item" >> $DATAFILE
+	free_incr=$(($base_mem_free-$item))
+	echo -n ", $free_incr" >> $DATAFILE
+
+	# Anon pages
+	item=$(cat /proc/meminfo | fgrep "AnonPages:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+
+	# Mapped pages
+	item=$(cat /proc/meminfo | fgrep "Mapped:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+
+	# Cached
+	item=$(cat /proc/meminfo | grep "^Cached:" | awk '{print $2}')
+	item=$(($item * 1024))
+	echo -n ", $item" >> $DATAFILE
+}
+
+function grab_stats() {
+	# If configured, dump the caches so we get a more stable
+	# view of what our static footprint really is
+	if (($DUMP_CACHES)) ; then
+		dump_caches
+	fi
+
+	# user space data
+		# USS taken by CC components
+	grab_cc_uss
+		# PSS taken all userspace
+	grab_cc_pss
+		# PSS taken all userspace
+	grab_all_pss
+		# PSS taken by dockerd
+	grab_dockerd_pss
+		# user as reported by smem
+	grab_user_smem
+
+	# System overview data
+		# System free and cached
+	grab_system
+
+	# kernel data
+		# The 'total kernel space taken' we can work out as:
+		# ktotal = ((free-avail)-user)
+		# So, we don't grab that number from smem, as that is what it does
+		# internally anyhow.
+		# Still try to grab any finer kernel details that we can though
+
+		# totals from slabinfo
+	grab_slab
+}
+
+function check_limits() {
+	mem_free=$(get_memfree)
+	if (($mem_free <= $MIN_MEMORY_FREE)); then
+		echo 1
+		return
+	fi
+
+	mem_consumed=$(($base_mem_avail-$mem_free))
+	if (($mem_consumed >= $MAX_MEMORY_CONSUMED)); then
+		echo 1
+		return
+	fi
+
+	echo 0
+}
+
+function go() {
+	for i in $(seq 1 $MAX_NUM_CONTAINERS); do
+		docker run -tid --runtime=$RUNTIME $PAYLOAD_RUNTIME_ARGS $PAYLOAD $PAYLOAD_ARGS
```

Comment:
Any good chance of having these vars expand in unfortunate ways? If so, could rework as arrays.

Reply:
so far they've been OK, but yes, it can get a little tricky if trying to pass, say, args to a runtime that does a `bash -c xxx` for instance. OK as is for now I think - we can fix if we run into a complex case we cannot handle...